optimum/exporters/ipex/modeling_utils.py

```diff
@@ -156,103 +253,335 @@ def _llama_model_forward(
+
+    input_lens = attention_mask.cumsum(-1)[:, -1].to(torch.int32)
+    if past_key_values is not None:
+        setattr(past_key_values, "input_lens", input_lens)
+
+    presents = None
+    all_self_attentions = () if output_attentions else None
+    all_cross_attentions = () if output_attentions and self.config.add_cross_attention else None
+    all_hidden_states = () if output_hidden_states else None
+    for i, block in enumerate(self.h):
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        outputs = block(
+            hidden_states,
+            layer_past=past_key_values,
+            attention_mask=attention_mask,
+            head_mask=head_mask[i],
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_attention_mask,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+        )
+
+        hidden_states = outputs[0]
+        if use_cache is True:
+            presents = outputs[1]
+
+        if output_attentions:
+            all_self_attentions = all_self_attentions + (outputs[2 if use_cache else 1],)
+            if self.config.add_cross_attention:
+                all_cross_attentions = all_cross_attentions + (outputs[3 if use_cache else 2],)
+
+    hidden_states = self.ln_f(hidden_states)
+    if hidden_states.shape[0] != batch_size * seq_length:
+        (hidden_states_copy.view(-1, hidden_states.shape[-1]))[attention_mask.view(-1) != 0] = hidden_states
+        hidden_states = hidden_states_copy
+
+    hidden_states = hidden_states.view(batch_size, -1, hidden_states.shape[-1])
+    # Add last hidden state
+    if output_hidden_states:
+        all_hidden_states = all_hidden_states + (hidden_states,)
+
+    if not return_dict:
+        return tuple(
+            v
+            for v in [hidden_states, presents, all_hidden_states, all_self_attentions, all_cross_attentions]
+            if v is not None
+        )
+
+    return BaseModelOutputWithPastAndCrossAttentions(
+        last_hidden_state=hidden_states,
+        past_key_values=presents,
+        hidden_states=all_hidden_states,
+        attentions=all_self_attentions,
+        cross_attentions=all_cross_attentions,
+    )
 
 
 class _IPEXAttention(nn.Module):
     def __init__(self, module, config) -> None:
         super().__init__()
         _setattr_from_module(self, module)
         self.config = config
-        self.ipex_scale_dot_product = IndirectAccessKVCacheAttention(text_max_length=config.max_position_embeddings)
-        if hasattr(config, "rope_theta"):
-            self.ipex_rope = RotaryEmbedding(
-                config.max_position_embeddings,
-                config.hidden_size // config.num_attention_heads,
-                config.rope_theta,
-                config.architectures[0],
-            )
+        self.module_device = next(module.parameters()).device.type
```

Comment:
do we wanna save only the device type here ?
`torch.xpu` device supports device ordinal like `torch.cuda`, wouldn't that create a problem with `self.kv_head_mapping` initialized on `xpu:0` by default ?

Reply:
This parameter is only for choosing the type of fused linear, so we only need to know the device type. Optimum-intel should not support cuda.

Reply:
It's just for selecting cpu/xpu linear to patch, we don't have any ops like `.to(self.module_device)`

Reply:
there's a tensor constructor call right after this line that uses `device=self.module_device`

Reply:
Fixed here: https://github.com/huggingface/optimum-intel/pull/1035